src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -2681,6 +2681,63 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
+        ASR::ttype_t *to_type = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+                                4, nullptr, 0));
+        ASR::expr_t *value = nullptr;
+        if (ASRUtils::is_character(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                char* c = ASR::down_cast<ASR::StringConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_s;
+                int64_t ival = std::string(c).size();
+                value = ASR::down_cast<ASR::expr_t>(make_IntegerConstant_t(al,
+                                loc, ival, to_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(
+                    ASR::make_StringLen_t(al, loc, arg, to_type, value));
+        } else if (ASR::is_a<ASR::Set_t>(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                int64_t ival = (int64_t)ASR::down_cast<ASR::SetConstant_t>(
+                                        ASRUtils::expr_value(arg))->n_elements;
+                value = ASR::down_cast<ASR::expr_t>(make_IntegerConstant_t(al,
+                                loc, ival, to_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(
+                    ASR::make_SetLen_t(al, loc, arg, to_type, value));
+        } else if (ASR::is_a<ASR::Tuple_t>(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                int64_t ival = (int64_t)ASR::down_cast<ASR::TupleConstant_t>(
+                                        ASRUtils::expr_value(arg))->n_elements;
+                value = ASR::down_cast<ASR::expr_t>(make_IntegerConstant_t(al,
+                                loc, ival, to_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(
+                    ASR::make_TupleLen_t(al, loc, arg, to_type, value));
+        } else if (ASR::is_a<ASR::List_t>(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                int64_t ival = (int64_t)ASR::down_cast<ASR::ListConstant_t>(
+                                        ASRUtils::expr_value(arg))->n_args;
+                value = ASR::down_cast<ASR::expr_t>(make_IntegerConstant_t(al,
+                                loc, ival, to_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(
+                    ASR::make_ListLen_t(al, loc, arg, to_type, value));
+        } else if (ASR::is_a<ASR::Dict_t>(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                int64_t ival = (int64_t)ASR::down_cast<ASR::DictConstant_t>(
+                                        ASRUtils::expr_value(arg))->n_keys;
+                value = ASR::down_cast<ASR::expr_t>(make_IntegerConstant_t(al,
+                                loc, ival, to_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(
+                    ASR::make_DictLen_t(al, loc, arg, to_type, value));
+        }
+        return nullptr;
```

Comment:
An error message would be better here. It should say that `len` is only supported for `Set`, `Character/String`, `Dict`, `List`. Returning `nullptr` now would create problems in future because then most probably we will receive a Segmentation fault while compiling a big library using Lpython. So, debugging process will involve first figuring out why segmentation fault is there and then getting to know that Ah!, the support for `len` isn't there for the input type. A nice error message would help in avoiding that hustle.

Reply:
Fixed in 593b8c448a62816d8ab323481a1918c18e70694f.

Reply:
Great. Thanks for doing this.